### PR TITLE
TensorFlow: abstract Mutex interfaces

### PR DESCRIPTION
**Pull request by compnerd**
Abstract out the mutex handling APIs to enable support for Windows'
threading primitives.  This enables mutexes on Windows via the SRWLOCK
(slim reader/writer locks).  The full set of the threading interfaces
is now fully isolated into Core/Threading.swift with cross-platform
implementations.

### CODE REVIEW

**Comment by compnerd** [comment]
CC: @dan-zheng @pschuh

File: Sources/TensorFlow/Core/Threading.swift [hunk 1]
```diff
@@ -146,6 +146,51 @@ class Thread {
     }
 }
 
+class Mutex {
```
Comment (dan-zheng) [inline Sources/TensorFlow/Core/Threading.swift]: Could you please add a simple documentation comment? Even though this is an internal API.

File: Sources/TensorFlow/Core/Runtime.swift [hunk 2]
```diff
@@ -404,13 +401,15 @@ internal extension _ExecutionContext {
...
-        let lockStatus = pthread_mutex_lock(&mutex)
+        let lockStatus = mutex.acquire()
         internalConsistencyCheck(lockStatus == 0)
         defer {
-            let unlockStatus = pthread_mutex_unlock(&mutex)
+            let unlockStatus = mutex.release()
             internalConsistencyCheck(unlockStatus == 0)
+#if !os(Windows)
             // Create a cancellation point.
             pthread_testcancel()
```
Comment (dan-zheng) [inline Sources/TensorFlow/Core/Runtime.swift]: Ignorant question: does it not make sense to move `pthread_testcancel()` to `Mutex.release`? It seems unfortunate that callers of `Mutex.release` still need an extra `#if !os(Windows)` directive with `pthread_testcancel()`.

Comment (compnerd) [reply Sources/TensorFlow/Core/Runtime.swift]: Im entirely happy to do that.  The reasoning behind not doing so immediately is that should every release be associated with a cancellation point?

File: Sources/TensorFlow/Core/Threading.swift (hunk 1 above)
Comment (compnerd) [reply Sources/TensorFlow/Core/Threading.swift]: Sure!

File: Sources/TensorFlow/Core/Runtime.swift (hunk 2 above)
Comment (dan-zheng) [reply Sources/TensorFlow/Core/Runtime.swift]: > The reasoning behind not doing so immediately is that should every release be associated with a cancellation point?

I'm clueless. Maybe @pschuh has an opinion?

Comment (compnerd) [reply Sources/TensorFlow/Core/Runtime.swift]: Okay, did a bit of code archeology with @pschuh ... I think that we came to the conclusion that we do not have a strong reason to keep this functionality at this time.  I can do a follow up change to remove the cancellation point support.

**Pull request merged**